optuna/cli.py

```diff
@@ -886,48 +896,184 @@ def take_action(self, parsed_args: Namespace) -> int:
+        command_name_to_subparser[command_name] = subparser
+
+    def _print_help(args: Namespace) -> None:
+        main_parser.print_help()
+
+    subparsers.add_parser("help", help="Show help message and exit.").set_defaults(
+        handler=_print_help
+    )
+    return main_parser, command_name_to_subparser
+
+
+def _get_parser(description: str = "") -> Tuple[ArgumentParser, Dict[str, ArgumentParser]]:
+    # Use `parent_parser` is necessary to avoid namespace conflict for -h/--help
+    # between `main_parser` and `subparser`.
+    parent_parser = ArgumentParser(add_help=False)
+    parent_parser = _add_common_arguments(parent_parser)
+
+    main_parser = ArgumentParser(description=description, parents=[parent_parser])
+    main_parser.add_argument(
+        "--version", action="version", version="{0} {1}".format("optuna", optuna.__version__)
+    )
+    main_parser, command_name_to_subparser = _add_commands(main_parser, parent_parser)
+    return main_parser, command_name_to_subparser
+
+
+def _preprocess_argv(argv: List[str]) -> Tuple[List[str], Tuple[str, List[str]]]:
+    # Some preprocess is necessary for argv because some subcommand includes space
+    # (e.g. optuna study optimize, optuna storage upgrade, ...)
+    argv = argv[1:] if len(argv) > 1 else ["help"]
+
+    # Some command consists of two strings.
+    command_candidate_to_indices: Dict[str, List[int]] = {}
+    for i in range(len(argv)):
+        command_candidate_to_indices[argv[i]] = [i]
+        for j in range(i + 1, len(argv)):
+            command_candidate_to_indices[" ".join([argv[i], argv[j]])] = [i, j]
+
+    command_indices = []
+    current_longest_command_length = 0
+    for command_candidate in command_candidate_to_indices:
+        for command_name in _COMMANDS:
+            command_length = len(command_name)
+            # Find the longest possible command.
+            if (
+                command_name == command_candidate
+                and command_length > current_longest_command_length
+            ):
+                command_indices = command_candidate_to_indices[command_candidate]
+                current_longest_command_length = command_length
 
-    def build_option_parser(
-        self, description: str, version: str, argparse_kwargs: Optional[Dict] = None
-    ) -> ArgumentParser:
+    preprocessed_command_name = " ".join(
+        [arg for i, arg in enumerate(argv) if i in command_indices]
+    )
+    options = [arg for i, arg in enumerate(argv) if i not in command_indices]
 
-        parser = super(_OptunaApp, self).build_option_parser(description, version, argparse_kwargs)
-        parser.add_argument("--storage", default=None, help="DB URL. (e.g. sqlite:///example.db)")
-        return parser
+    preprocessed_argv = [preprocessed_command_name] + options
+    preprocessed_argv = [arg for arg in preprocessed_argv if arg != ""]
+    return preprocessed_argv, (preprocessed_command_name, options)
```

Comment:
[Question] Where are these operations processed in the precious implementation?

Reply:
This code is re-implementation of `find_command` function in https://github.com/openstack/cliff/blob/master/cliff/commandmanager.py .

Reply:
And currently I fixed code of this part for more readability.